In [8]:
#import libraries
import pandas as pd
import numpy as np
import os

In [9]:
#create path
path = r'C:\Users\JocGa\Instacart Basket Analysis'

In [10]:
#import dataframe
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_update.pkl'))

In [11]:
#overview of dataframe
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busys           Most orders  
1  Regularly busy     Slowest days        Average orders  
2  Regularly busy     Slowest days           Most orders  
3      Least busy     Slowest days        Average orders  
4      Least busy     Slowest days           Most orders

In [12]:
#number of rows and columns
df_ords_prods_merge.shape

(32404859, 18)

In [13]:
# Importing data set departments.csv
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [14]:
# Transposing df_dep horizontal view
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#Step 2 In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [15]:
#finding the aggregated mean for the entrie dataframe
df_ords_prods_merge.groupby('product_name')

In [16]:
#performing single statisitc on 'order_number' by dept
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#observations all departments are showen along with order_number mean; department 10 (bulk) has a mean around 20, while 5 (alcohol) has a mean around 15.2. that dept bulk sells considerably more on average than dept alcohol. 

#Step 4 Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [17]:
#creating the "max_order" column
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [18]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   
5                    19.0         196                  1          1   both   
6                    20.0         196                  1          1   both   
7                    14.0         196                  2          1   both   
8                     0.0         196                  4          1   both   
9                    30.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   
5         Soda        77              7     9.0  Mid-range product   
6         Soda        77              7     9.0  Mid-range product   
7         Soda        77              7     9.0  Mid-range product   
8         Soda        77              7     9.0  Mid-range product   
9         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Regularly busys           Most orders         10  
1  Regularly busy     Slowest days        Average orders         10  
2  Regularly busy     Slowest days           Most orders         10  
3      Least busy     Slowest days        Average orders         10  
4      Least busy     Slowest days           Most orders         10  
5  Regularly busy  Regularly busys        Average orders         10  
6  Regularly busy     Busiest days           Most orders         10  
7  Regularly busy     Busiest days           Most orders         10  
8  Regularly busy     Busiest days           Most orders         10  
9      Least busy     Slowest days           Most orders         10

#each value in column max_order corresponds to the maximum number of orders made by each uesr ID

In [19]:
# Creating a 'loyalty' flag
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [20]:
# Creating a 'loyalty' flag
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [21]:
#creating a 'loyalty' flag
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [22]:
#checking dataset
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

#observations majority of customers fall into the "regular customer" category

In [23]:
#checking columns user id, loyalty_flag, and order_number
df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(25)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2

#Step 5 The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [24]:
#performing multiple aggregations
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'max', 'min', 'sum']})

prices                         
                      mean   max  min           sum
loyalty_flag                                       
Loyal customer    7.772831  25.0  1.0  7.993651e+07
New customer      7.800320  25.0  1.0  4.870512e+07
Regular customer  7.797431  25.0  1.0  1.237981e+08

#observations the avarage of prices spend among three groups are the same, but the sum spend among the three groups shows that loyal customer spend more.

#Step 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”



In [25]:
#creating the "average price" column
df_ords_prods_merge['average_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [26]:
#overview of dataframe
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... aisle_id  department_id  prices    price_range_loc     busiest_day  \
0  ...       77              7     9.0  Mid-range product  Regularly busy   
1  ...       77              7     9.0  Mid-range product  Regularly busy   
2  ...       77              7     9.0  Mid-range product  Regularly busy   
3  ...       77              7     9.0  Mid-range product      Least busy   
4  ...       77              7     9.0  Mid-range product      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag average_price  
0  Regularly busys           Most orders        10  New customer      6.367797  
1     Slowest days        Average orders        10  New customer      6.367797  
2     Slowest days           Most orders        10  New customer      6.367797  
3     Slowest days        Average orders        10  New customer      6.367797  
4     Slowest days           Most orders        10  New customer      6.367797  

[5 rows x 21 columns]

In [27]:
# Creating a 'spending_flag' flag
df_ords_prods_merge.loc[df_ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [28]:
# Creating a 'spending_flag' flag
df_ords_prods_merge.loc[df_ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [29]:
#check counts
df_ords_prods_merge['spending_flag'].value_counts()

Low spender     32285165
High spender      119694
Name: spending_flag, dtype: int64

#observations there are more low spender purchasing low avarage prices items 

#Step 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

    If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
    If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
    If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”



In [30]:
#creating the "order_frequency_flag" column
df_ords_prods_merge['median_prior_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [31]:
#overview of dataframe
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... prices    price_range_loc     busiest_day     busiest_days  \
0  ...    9.0  Mid-range product  Regularly busy  Regularly busys   
1  ...    9.0  Mid-range product  Regularly busy     Slowest days   
2  ...    9.0  Mid-range product  Regularly busy     Slowest days   
3  ...    9.0  Mid-range product      Least busy     Slowest days   
4  ...    9.0  Mid-range product      Least busy     Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_price  spending_flag  \
0           Most orders        10  New customer      6.367797    Low spender   
1        Average orders        10  New customer      6.367797    Low spender   
2           Most orders        10  New customer      6.367797    Low spender   
3        Average orders        10  New customer      6.367797    Low spender   
4           Most orders        10  New customer      6.367797    Low spender   

  median_prior_orders  
0                20.5  
1                20.5  
2                20.5  
3                20.5  
4                20.5  

[5 rows x 23 columns]

In [32]:
# Creating a 'order_frequency_flag' flag
df_ords_prods_merge.loc[df_ords_prods_merge['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [33]:
# Creating a 'order_frequency_flag' flag
df_ords_prods_merge.loc[(df_ords_prods_merge['median_prior_orders'] > 10) & (df_ords_prods_merge['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [34]:
# Creating a 'order_frequency_flag' flag
df_ords_prods_merge.loc[df_ords_prods_merge['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [35]:
#overview of dataframe
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ...    price_range_loc     busiest_day     busiest_days  \
0  ...  Mid-range product  Regularly busy  Regularly busys   
1  ...  Mid-range product  Regularly busy     Slowest days   
2  ...  Mid-range product  Regularly busy     Slowest days   
3  ...  Mid-range product      Least busy     Slowest days   
4  ...  Mid-range product      Least busy     Slowest days   

   busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0            Most orders        10  New customer      6.367797   Low spender   
1         Average orders        10  New customer      6.367797   Low spender   
2            Most orders        10  New customer      6.367797   Low spender   
3         Average orders        10  New customer      6.367797   Low spender   
4            Most orders        10  New customer      6.367797   Low spender   

   median_prior_orders   order_frequency_flag  
0                 20.5  Non-frequent customer  
1                 20.5  Non-frequent customer  
2                 20.5  Non-frequent customer  
3                 20.5  Non-frequent customer  
4                 20.5  Non-frequent customer  

[5 rows x 24 columns]

In [36]:
#check counts
df_ords_prods_merge['order_frequency_flag'].value_counts()

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
Name: order_frequency_flag, dtype: int64

#observations there are more frequent customer than non-frequent customer based on 'days since prior order'

In [37]:
#number of columns and rows
df_ords_prods_merge.shape

(32404859, 24)

# Export data

In [38]:
#export dataframe
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_update2.pkl'))